python/tvm/driver/tvmc/extensions.py

```diff
@@ -0,0 +1,128 @@
+
+
+def get_extensions():
+    """Returns all loaded extensions."""
+
+    for ext in _EXTENSIONS:
+        yield ext
+
+
+def load_extensions(paths):
+    """
+    Loads extensions from the given locations.
+
+    Extensions must implement the `TVMExtension` interface and be stored in a directory called
+    `tvm_extension`.
+    """
+
+    path_backup = copy.copy(sys.path)
+    sys.path.extend(paths)
+
+    top_modules = []
+    try:
+        mod = importlib.import_module("tvm_extension")
+        top_modules.append(mod)
+    except ImportError:
+        pass
+
+    sys.path.clear()
+    sys.path.extend(path_backup)
+
+    extension_classes = _scan_all(top_modules)
+    for ext_cls in extension_classes:
+        _EXTENSIONS.append(ext_cls())
+
+
+def _scan_all(top_level):
+    scanned_extensions = []
+    for mdl in top_level:
+        for importer, modname, _ in pkgutil.walk_packages(
+            path=mdl.__path__, prefix=mdl.__name__ + ".", onerror=lambda x: None
+        ):
+            try:
+                module_name = modname.rsplit(".", 1)[-1]
+                # If module's name starts with "_", do not load the module.
+                # But if the module's name starts with a "__", then load the
+                # module.
+                if module_name.startswith("_") and not module_name.startswith("__"):
+                    continue
+
+                with warnings.catch_warnings(record=True) as recorded_warnings:
+                    if sys.version_info < (3, 10):
+                        m = importer.find_module(modname)  # type: ignore
+                        assert m is not None
+                        loaded_mod = m.load_module(modname)
+                    else:
+                        spec = importer.find_spec(modname)
+                        assert spec is not None
+                        if modname in sys.modules:
+                            loaded_mod = sys.modules[modname]
+                        else:
+                            loaded_mod = importlib.util.module_from_spec(spec)
+                        if loaded_mod is not None:
+                            spec.loader.exec_module(loaded_mod)
+                            sys.modules[modname] = loaded_mod
+
+                if len(recorded_warnings) > 0:
+                    for warning in recorded_warnings:
+                        warnings.showwarning(
+                            message=warning.message,
+                            category=warning.category,
+                            filename=warning.filename,
+                            lineno=warning.lineno,
+                            file=warning.file,
+                            line=warning.line,
+                        )
+
+                if loaded_mod is not None:
+                    for _name, obj in inspect.getmembers(loaded_mod):
+                        if _is_concrete_extension_type(obj):
+                            scanned_extensions.append(obj)
+            except ImportError as err:
```

Comment:
would be good to add a test for this case

Reply:
I tried adding this, but encountered an issue with race conditions. The way the `sys.path` is currently modified by extensions.py is problematic when more than one test is run in parallel. Is there already an approach for serializing tests in TVM? There doesn't seem to be an upstream solution: https://github.com/pytest-dev/pytest-xdist/issues/18

Reply:
Could we instead avoid modifying `sys.path` when loading a module? I had a naive search and wondered if [this]( https://stackoverflow.com/questions/67631/how-can-i-import-a-module-dynamically-given-the-full-path ) could help? cc @Mousius @leandron who might know more about these things

Reply:
While such an approach should work for single source files, I'm not sure how that would work for packages or namespace packages.

Reply:
Apologies for the delay, I had a think about this but wasn't able to come up with a better solution myself, so I won't block on this